Add documentation builder

rsdy opened the pull request:
Build and deploy documentation to foniod.org. Normally this should only happen after releases.

rhdxmr commented:
Hello @rsdy 
Long time no see.

I reviewed your PR and I found one error when I run `cargo doc` with `--all-features` option.

See the error message below:
```
  --> cargo-bpf/src/build_c.rs:10:20   |10 | use crate::build::{get_llc_executable, kernel_headers, BUILD_FLAGS};   |                    ^^^^^^^^^^^^^^^^^^  ^^^^^^^^^^^^^^  ^^^^^^^^^^^ no `BUILD_FLAGS` in `build`
   |                    |                   |
   |                    |                   no `kernel_headers` in `build`
   |                    no `get_llc_executable` in `build`

error: Compilation failed, aborting rustdoc

error[E0432]: unresolved imports `crate::build::get_llc_executable`, `crate::build::kernel_headers`, `crate::build::BUILD_FLAGS`
  --> cargo-bpf/src/build_c.rs:10:20
   |
10 | use crate::build::{get_llc_executable, kernel_headers, BUILD_FLAGS};
   |                    ^^^^^^^^^^^^^^^^^^  ^^^^^^^^^^^^^^  ^^^^^^^^^^^ no `BUILD_FLAGS` in `build`
   |                    |                   |
   |                    |                   no `kernel_headers` in `build`
   |                    no `get_llc_executable` in `build`

```


`cargo-bpf/src/build_c.rs` seems to be not used for a very long time. I don't know since when this module is not able to be compiled properly. Hmm..

And I am not sure what `build_c.rs` is for. So IMO it would be better to remove this code because it is dead code.

----

I have one question. Your PR only contains `git push` at the end. I don't understand how this command changes the content of foniod.org. Can you explain me this? Thanks in advance.

rsdy commented:
@rhdxmr Sorry for the radio silence, I've been busy on other projects.

1. `build_c.rs` is responsible for building C code. I'm inclined to fix it, since it's a good tool for interfacing C code with Rust, and it seems like the issue is with LLVM.

2. `git push` at the end is supposed to update the repo. If you look at the following snippet, the work happens in the checked out directory that is https://github.com/foniod/foniod.org 

        # check out the website repo
        git clone https://$GITHUB_TOKEN:@github.com/foniod/foniod.org publish
        
        # clean up the existing docs
        rm -rf publish/api/
        mkdir -p publish/api/
        cp -r target/doc/* publish/api/

        # this is an automatic redirect that btw needs to be fixed ;)
        >publish/api/index.html echo '<meta http-equiv="refresh" content="0; url=https://ingraind.org/api/redbpf/" />'

        # change directory, commit changeset, push
        cd publish
        git add *
        git commit -a -m '[CI] Update redbpf docs'
        git push

rhdxmr commented:
@rsdy 
I don't understand how `git push` deploys the content to foniod.org site. It just pushes the code to github.

If deploying the content to foniod.org is completed by `git push`, the foniod.org site should already serve API docs correctly. Because the repo already has docs.

rsdy commented:
@rhdxmr Sorry, you're right. I think that's a separate bug, because I have changed the location of the build artifacts from root to `/docs/`. Will fix that on this branch once doc build goes through.

On another note, as you see, I removed `build_c.rs`. I took another look, and it seems superflous at this stage.

rhdxmr commented:
@rsdy unfortunately the build test of documentation had failed because of an authentication problem.

BTW what projects are you working on these days? Are there some interesting subjects or new tech? I am just curious. Can you tell me if you don't mind?

rsdy commented:
@caniszczyk do you think you can help us set up an api.foniod.org domain on netlify to serve static build content from this repo? I've tried but it doesn't seem like I have the necessary permissions.

caniszczyk commented:
I created https://app.netlify.com/sites/foniodapi/overview which you should
have access to tweak the Netilfy build/settings

See if that works, you may need to tweak the netlify build config based on
what you need.

Is the repo https://github.com/foniod/foniod.org or are you using something
for api.foniod.org (e.g., what github repo is the source/build)

On Sun, Sep 19, 2021 at 7:58 PM rsdy ***@***.***> wrote:

> @caniszczyk <https://github.com/caniszczyk> do you think you can help us
> set up an api.foniod.org domain on netlify to serve static build content
> from this repo? I've tried but it doesn't seem like I have the necessary
> permissions.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/foniod/redbpf/pull/186#issuecomment-922512410>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAPSIPPWRHZQTK5YJGKOULUCYQD3ANCNFSM5EKL3WOA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Cheers,

Chris Aniszczyk
https://aniszczyk.org

rsdy commented:
@caniszczyk Thanks for this. We're looking to create a special branch on this repo that hosts documentation. This is the cleanest way to host the API docs due to permission boundaries in GitHub. Will test this some more, but I should be set with the required permissions on both ends.

Please also delete the following sites on Netlify, they're no longer needed:

 * https://app.netlify.com/sites/vigorous-jones-71b45d
 * https://app.netlify.com/sites/naughty-bose-b95ee1

caniszczyk commented:
np, deleted the sites

On Fri, Sep 24, 2021 at 6:47 AM rsdy ***@***.***> wrote:

> @caniszczyk <https://github.com/caniszczyk> Thanks for this. We're
> looking to create a special branch on this repo that hosts documentation.
> This is the cleanest way to host the API docs due to permission boundaries
> in GitHub. Will test this some more, but I should be set with the required
> permissions on both ends.
>
> Please also delete the following sites on Netlify, they're no longer
> needed:
>
>    - https://app.netlify.com/sites/vigorous-jones-71b45d
>    - https://app.netlify.com/sites/naughty-bose-b95ee1
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/foniod/redbpf/pull/186#issuecomment-926530604>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAPSILQPFQQTKXJCE3E6Y3UDRJNZANCNFSM5EKL3WOA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Cheers,

Chris Aniszczyk
https://aniszczyk.org

rhdxmr commented:
I just re-ran all jobs of github action and I will see whether the building documentation test succeeds

rsdy commented:
@rhdxmr I haven't followed up with the necessary changes to make this build successfully. I'll be able to take a look one of these days.

rsdy commented:
closing as superceded by #225
